### PR TITLE
External user login confirm

### LINKED ISSUE #1393: Add confirmation for login with login key
External users can login by clicking on the login URL sent to them via email. Before users are actually authenticated on the platform after clicking on this link, they should have to confirm the login by, e.g., clicking a confirmation button in a modal or on a designated page.

This prevents cases in which the URL is already requested for, e.g., a preview of the page and thus invalidates the login key which results in a new link being sent when users actively click on the link (which is then already the second GET request).

### PR DESCRIPTION
**Pull request by niklasmohrin**
Fixes #1393

### CODE REVIEW

File: evap/evaluation/templates/external_user_confirm_login.html [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+                <h4 class="card-title text-primary">
+                    {% trans 'Confirm your login' %}
+                </h4>
+                <p class="card-text">
+                    {% trans 'By clicking this button, you will be logged in and your login link will be invalidated.' %}
+                </p>
+                <form class="d-flex flex-column" role="form" method="POST">
+                    {% csrf_token %}
+                    <div class="d-flex justify-content-center">
+                        <button type="submit" class="btn btn-primary login-button">{% trans 'Login as' %} {{ username }}</button>
```
Comment (janno42) [inline evap/evaluation/templates/external_user_confirm_login.html]: can you please make the username part of the translation? we probably want to translate it as `Als {{ username }} anmelden` in german.

File: evap/evaluation/tests/test_auth.py [hunk 2]
```diff
@@ -30,25 +30,29 @@ def test_login_url_generation(self):
         self.assertEqual(reversed_url, '/key/{}'.format(self.external_user.login_key))
 
     def test_login_url_works(self):
-        self.assertRedirects(self.app.get(reverse("contributor:index")), "/?next=/contributor/")
```
Comment (He3lixxx) [inline evap/evaluation/tests/test_auth.py]: Why did you remove this?

File: evap/evaluation/tests/test_auth.py [hunk 3]
```diff
@@ -30,25 +30,29 @@ def test_login_url_generation(self):
...
+        self.assertContains(page, 'Login')
+        self.assertContains(page, self.external_user.full_name)
+
+        page = self.app.post(url_with_key).follow().follow()
         self.assertContains(page, 'Logout')
+        self.assertContains(page, self.external_user.full_name)
 
     def test_login_key_valid_only_once(self):
-        page = self.app.get(reverse("evaluation:login_key_authentication", args=[self.external_user.login_key])).follow().follow()
-        self.assertContains(page, self.external_user.full_name)
```
Comment (He3lixxx) [inline evap/evaluation/tests/test_auth.py]: I'd keep this here (looks like you moved it to the other test? I guess both tests should have it)

File: evap/evaluation/tests/test_auth.py [hunk 4]
```diff
@@ -30,25 +30,29 @@ def test_login_url_generation(self):
...
     def test_login_url_works(self):
-        self.assertRedirects(self.app.get(reverse("contributor:index")), "/?next=/contributor/")
-
         url_with_key = reverse("evaluation:login_key_authentication", args=[self.external_user.login_key])
-        page = self.app.get(url_with_key).follow().follow()
+        page = self.app.get(url_with_key)
+        self.assertContains(page, 'Login')
+        self.assertContains(page, self.external_user.full_name)
+
+        page = self.app.post(url_with_key).follow().follow()
```
Comment (He3lixxx) [inline evap/evaluation/tests/test_auth.py]: While this tests the typical login flow, it does not test whether the special case we are doing this for works (email client issues get, browser issues get, browser issues post).
I'd add
```
old_key = self.external_user.login_key
old_valid_until = self.external_user.login_key_valid_until
```
before the get and
```
self.external_user.refresh_from_db()
self.assertEqual(old_login_key, self.external_user.login_key)
self.assertEqual(old_login_key_valid_until, self.external_user.login_key_valid_until)
```
afterwards.

Alternatively, we could add another `self.app.get` between the `get` and the `post` calls, however, I think this might work just because the app remembers something about the user. Not sure about this, though (how much state does `self.app` have in the tests?)

File: evap/evaluation/templates/external_user_confirm_login.html [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+                <h4 class="card-title text-primary">
+                    {% trans 'Confirm your login' %}
+                </h4>
+                <p class="card-text">
+                    {% trans 'By clicking this button, you will be logged in and your login link will be invalidated.' %}
+                </p>
+                <form class="d-flex flex-column" role="form" method="POST">
+                    {% csrf_token %}
+                    <div class="d-flex justify-content-center">
+                        <button type="submit" class="btn btn-primary login-button">{% blocktrans %}Login as {{ username }} {% endblocktrans %}</button>
```
Comment (janno42) [inline evap/evaluation/templates/external_user_confirm_login.html]: there is a trailing space in the string which shouldn't be there ;)

**Comment by janno42** [comment]
thanks!
